### DISCUSSION TITLE
leiden algorithm run time is much longer than the louvain algorithm

### DISCUSSION THREAD

**Discussion by Liripo** [satijalab/seurat#6754] [Q&A]
Leiden algorithm run 16 mins,but louvain algorithm only run 30s.
```R
> bench::bench_time({
   object <- FindClusters(object,
      algorithm = 4,
      method = "igraph",
      resolution = 0.8
  )
 })
process    real 
  1.16m  16.64m 
```
louvain algorithm
```R
> bench::bench_time({
   object <- FindClusters(object,
    algorithm = 1,
    method = "igraph",
    resolution = 0.8
   )
 })
process    real 
  6.21s  30.26s
```

And my object has 40,000 cells.  Is this normal？

**Comment by alanocallaghan** [comment]
This isn't really an expected result, see for example https://www.nature.com/articles/s41598-019-41695-z/
where Leiden is found to be much faster than Louvain.

It may be because Louvain is implemented directly in C++ (here: https://github.com/satijalab/seurat/blob/4e868fcde49dc0a3df47f94f5fb54a421bfdf7bc/src/ModularityOptimizer.cpp)
while Leiden uses a python module via reticulate (here: https://github.com/satijalab/seurat/blob/c4638730d0639d770ad12c35f50d19108e0491db/R/clustering.R#L1627-L1716)

**Reply by samuel-marsh** [reply]
Just chiming in as note I have also experienced this and echoing @Alanocallaghan that was my guess as well since Seurat implementation calls Leiden package which sends things to reticulate and into python.

Leiden authors note there are a couple packages which include some part of python leidenalg functionality but none are complete (https://github.com/vtraag/leidenalg#usage).

It would appear though that Cole Trapnell’s lab implementation (https://cran.rstudio.com/web/packages/leidenbase/index.html) might get closest for speed as it aims to implement in C/C++.  I haven’t dug entirely into functionality and whether or not that could be used here but thought I’d mention.

Best,
Sam

P.S. @mojaveazure hope you don’t mind but tagging here to see if Seurat team has experienced this or discussed potential changes?

**Comment by mojaveazure** [comment]
This slowdown is likely because we're using Leiden/reticulate to call the Python implementation of leiden rather than using an R-native implementation. We haven't looked into any of the R/C++ implementations as
- we last looked at this in early 2021
- all currently R-native leiden implementations are relatively new
- we require any and all dependencies, including suggested ones, to be on CRAN or Bioconductor
- we just don't have the bandwidth to continually evaluate new clustering methods

If anyone would like to investigate whether or not Leidenbase or leidenAlg (Kharchenko lab) are indeed faster than Leiden/reticulate and make a PR to support one of those R-native packages, we would be happy to include it. However, our focus at this time is on other priorities

**Reply by samuel-marsh** [reply]
Thanks so much for insight and totally understand from dev perspective.  Don’t know that I have bandwidth to look into PR either but maybe some time in new year I’ll come back around unless someone else does first.

Thanks again!
Sam

**Reply by alanocallaghan** [reply]
They're both substantially faster, but the Trapnell group one (`leidenbase`) seems faster and much better documented

```r
library("leidenbase")
library("leidenAlg")
library("leiden")
library("igraph")

fpath <- system.file('testdata', 'igraph_n1500_edgelist.txt.gz', package='leidenbase')
zfp <- gzfile(fpath)
igraph <- read_graph(file = zfp, format='edgelist')

microbenchmark::microbenchmark(leiden.community(igraph), leiden_find_partition(igraph), leiden(igraph))
# Unit: milliseconds
#                           expr         min          lq        mean      median
#       leiden.community(igraph)    52.98505    55.16530    63.16430    55.95267
#  leiden_find_partition(igraph)    29.22238    30.54163    39.15174    37.57693
#                 leiden(igraph) 17140.52418 20671.12398 21172.11517 21128.94100
#           uq         max neval cld
#     75.38524    80.10432    10  a 
#     40.30238    56.91094    10  a 
#  22268.67866 24597.89747    10   b
```
